docs/contributing/999-rfcs/2022-12-08-wing-inputs.md

```diff
@@ -0,0 +1,202 @@
+---
+title: "#2726 Wing Inputs"
+description: "How to create, share, manage inputs"
+---
+
+# Wing Inputs
+- **Author(s)**: @Hasanaburayyan
+- **Submission Date**: {2023-12-08}
+- **Stage**: Draft
+
+## Background
+
+An input is a value used by either an application or platform that is not necessarily defined within the application's source code.
+
+Inputs can be Platform agnostic, such as an a retry count that is dictated by business requirements. Other inputs may be Platform specific, like a DNS configurations.
+
+### Context
+
+There are a few things that are considered out-of-scope for this RFC. 
+- Inputs are all treated as string data, but helper methods can convert them to primitive Wing types like num, bool, and Json
+- Inputs cannot be created, or updated inflight
+
+## Use cases
+
+The following table contains some use cases and is by no means exhaustive. 
+
+| **Use case** | **Secure** | Consumed |
+| ---- | ---- | ---- |
+| Connection strings | Yes | Inflight |
+| SSH Keys | Yes | Inflight |
+| Bucket count | No | Both |
+| Connection strings | Yes | Inflight |
+| SSH Keys | Yes | Inflight |
+| Bucket count | No | Both |
+| Retry counts | No | Inflight |
+| API Keys | Yes | Inflight |
+| DNS information (certificates, hosted zones) | No | Preflight |
+| Autoscaling parameters | No | Preflight |
+| Feature Flags | No | Both |
+| Platform Input (state-file name, etc..) | No | Preflight |
+
+
+## Configuring and Managing Inputs
+
+### The Wing Configuration File
+
+Inputs are defined via a Wing configuration file. These configuration files are called `wing.toml` and support other file formats such as `json` and `yaml`. 
+
+For instance in order to create a single input that will be used to determine how many buckets are created in our preflight code, we can do the following:
+
+```toml
+# wing.toml
+bucketCount = 3
+```
+
+```js
+// main.w
+bring cloud;
+let bucketCount = cloud.Input("bucketCount");
+
+let buckets = MutArray<cloud.Bucket>[];
+
+for i in 0..buckets.asNum() {
+  buckets.push(new cloud.Bucket() as "bucket_${i}");
+}
+```
+
+
+Additionally lets imagine we wanted to create resources based on feature flags provided at compile time.
+
+```toml
+email-list = "one@email.com, two@gmail.com"
+
+[ features ]
+notifications = false
+```
+
+```js
+let enableNotifications = cloud.Input("/features/notifications");
+let emailList = cloud.Input("/email-list"); // comma separated string
+
+let inbox = new cloud.Bucket() as "inbox";
+
+// conditionally provision infrastructure based on enableNotification
+if enableNotifications.asBool() {
+	inbox.onCreate(inflight () => {
+		// use emailList input inflight
+		for email in emailList.asStr().split(",") {
```

Comment:
you need to trim the email in this example because there are spaces in the toml file

```
email-list = "one@email.com, two@gmail.com"
```

so the second email will include space